### PR TITLE
osd: break down osd usage into data, omap, metadata buckets

### PR DESCRIPTION
**Pull request by liewegas**
Rocksdb can tell us how much space (approximately) is consumed by omap.
And BlueStore knows how much it has expended on data.  Use this to report up
all the way to 'cpeh osd df'.

Note that there is some weirdness here because the 'db' and 'wal' devices aren't
included in the total for osd size, so things don't quite add up.  If we want to
address that too, we also need to separate out (approximate) freespace for data vs
omap probably?

I expect in most cases that will be in the noise, and this is certainly a step forward,
but it does have teh potential to confuse users.

For example, with bluestore, after some rados bench with --write-omap we get
```
ID CLASS WEIGHT  REWEIGHT SIZE  USE   DATA OMAP  META AVAIL %USE  VAR  PGS 
 0   ssd 0.00980  1.00000 10.1G 4.74G 288K 4.47G 657M 5.33G 47.07 1.00   8 
                    TOTAL 10.1G 4.74G 288K 4.47G 657M 5.33G 47.07          
```
With filestore we can't tell how much is metadata overhead (either in rocksdb/leveldb or in XFS), so META is always zero:
```
ID CLASS WEIGHT  REWEIGHT SIZE USE  DATA  OMAP META AVAIL %USE  VAR  PGS 
 0   ssd 0.36369  1.00000 372G 299G 73.7G 112M    0 73.6G 80.25 1.00   8 
                    TOTAL 372G 299G 73.7G 112M    0 73.6G 80.25          
```

### CODE REVIEW

**Comment by yuyuyu101** [comment]
to the end user like rgw/rbd, omap is the same as metadata I think. I guess omap and metadata separation is only helpful for developer or specialist.

Anyway, it's ok to expose omap usage.

File: src/kv/KeyValueDB.h [hunk 1]
```diff
@@ -15,18 +15,31 @@
...
+    ColumnFamily(const string &name, const string &option)
+      : name(name), option(option) {}
+  };
+
   class TransactionImpl {
   public:
     /// Set Keys
     void set(
-      const std::string &prefix,                 ///< [in] Prefix for keys
+      const std::string &prefix,                      ///< [in] Prefix for keys, or CF name
```
Comment (amitkumar50) [inline src/kv/KeyValueDB.h]: tab left

**Comment by liewegas** [comment]
@ifed01 thoughts on this? It conflicts with the other per-pool piece, but it's smaller and doesn't involve a disk format change (and it's more or less ready).  The main problem right now is that it triggers an uninitialized memory access in rocksdb from valgrind; I opened https://github.com/facebook/rocksdb/issues/3707.

We could skip the metadata accounting (or hide it perhaps in 'osd df'), since, notably, filestore doesn't know how to populate that field (it calls all of the leveldb/rocksdb 'omap' and can't tell how much space XFS is spending on internal metadata vs actual data).

**Comment by ifed01** [comment]
@liewegas - once technically this is OK I have some other concerns. 
I'm not sure regular users are interested in that OMAP/metadata stuff at all. It looks like this can't result in any administrating/tuning decisions. Neither can warn about something. Or I missed something?

What they are rather in need is (mostly for BlueStore) more detailed report on both block and db volume stats(used/free/total plus BlueFS used/total at block device if DB vol != Block one). As well as clearer view on allocation overhead, compression benefits etc. This information will help to make some decisions and indicate some performance issues (e.g. when bluefs starts to use slow device).

**Comment by liewegas** [comment]
@ifed01 I agree we need to measure more here... I see this as just step one.  There's also the by-pool dimension that your other PR addresses (which I think can come next).  Then also more insight into the db vs main split.  Maybe the "available" stats can break down by data and omap?  And we need to surface (somewhere!) whether and when spillover is/will happen.

**Comment by liewegas** [comment]
@ifed01 I think I've addressed your comments.. can you take a look at the top commits before I squash?  Thanks!

File: src/os/filestore/DBObjectMap.h [hunk 2]
```diff
@@ -59,6 +59,8 @@ class DBObjectMap : public ObjectMap {
 public:
   boost::scoped_ptr<KeyValueDB> db;
 
+  KeyValueDB *get_db() { return db.get(); }
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/filestore/DBObjectMap.h]: override missed

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -6893,13 +6893,25 @@ int BlueStore::statfs(struct store_statfs_t *buf)
...
+
   uint64_t thin_total, thin_avail;
   if (bdev->get_thin_utilization(&thin_total, &thin_avail)) {
     buf->total += thin_total;
 
     // we are limited by both the size of the virtual device and the
     // underlying physical device.
     bfree = std::min(bfree, thin_avail);
+
+    std::lock_guard<std::mutex> l(vstatfs_lock);
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: no need for the lock

**Pull request merged**